inference-engine/src/transformations/src/transformations/common_optimizations/mvn_fusion.cpp

```diff
@@ -0,0 +1,339 @@
+#include "transformations/common_optimizations/mvn_fusion.hpp"
+#include "transformations/utils/utils.hpp"
+
+#include <memory>
+#include <vector>
+
+#include <ngraph/opsets/opset6.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/pattern/op/or.hpp>
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::MVNFusion, "MVNFusion", 0);
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::MVNFusionOutsideSqrt, "MVNFusionOutsideSqrt", 0);
+
+ngraph::pass::MVNFusionOutsideSqrt::MVNFusionOutsideSqrt() {
+    MATCHER_SCOPE(MVNFusionOutsideSqrt);
+    auto single_consumer = pattern::consumers_count(1);
+
+    // Detect MVN decomposition pattern:
+    // (x - ReduceMean(x, axes)) / (Sqrt(ReduceMean((x - ReduceMean(x, axes)) ^ 2)) + eps)
+    auto x = pattern::any_input();
+
+    // (x - ReduceMean(x, axes))
+    //     `------mean1-------'
+    auto mean1_axes = pattern::wrap_type<opset6::Constant>();
+    auto mean1 = pattern::wrap_type<opset6::ReduceMean>({ x, mean1_axes }, single_consumer);
+
+    // (x - ReduceMean(x, axes))
+    // `-sub1------------------'
+    auto sub1 = pattern::wrap_type<opset6::Subtract>({ x, mean1 });
+
+    // Sqrt(ReduceMean((x - ReduceMean(x, axes)) ^ 2))
+    //                     `---mean2----------'
+    auto mean2_axes = pattern::wrap_type<opset6::Constant>();
+    auto mean2 = pattern::wrap_type<opset6::ReduceMean>({ x, mean2_axes }, single_consumer);
+
+    // Sqrt(ReduceMean((x - ReduceMean(x, axes)) ^ 2))
+    //                 `-sub2------------------'
+    auto sub2 = pattern::wrap_type<opset6::Subtract>({ x, mean2 }, single_consumer);
+
+    const auto reuseSub1OrNot = std::make_shared<pattern::op::Or>(OutputVector{ sub1, sub2 });
+
+    auto cast = pattern::wrap_type<opset6::Convert>({ reuseSub1OrNot }, single_consumer);
+    const auto hasConvertOrNot = std::make_shared<pattern::op::Or>(OutputVector{ cast, reuseSub1OrNot });
+
+    // Sqrt(ReduceMean((x - ReduceMean(x, axes)) ^ 2))
+    //                 `---------------------power--'
+    auto const_2 = pattern::wrap_type<opset6::Constant>();
+    auto power = pattern::wrap_type<opset6::Power>({ hasConvertOrNot, const_2 }, single_consumer);
+
+    // Sqrt(ReduceMean((x - ReduceMean(x, axes)) ^ 2))
+    //     `---mean3--------------------------------'
+    auto mean3_axes = pattern::wrap_type<opset6::Constant>();
+    auto mean3 = pattern::wrap_type<opset6::ReduceMean>({ power, mean3_axes }, single_consumer);
+
+    // Sqrt(ReduceMean((x - ReduceMean(x, axes)) ^ 2))
+    // `--Power--------------------------------------'
+    auto const_0_5 = pattern::wrap_type<ngraph::opset6::Constant>();
+    auto power_sqrt = pattern::wrap_type<opset6::Power>({ mean3, const_0_5 }, single_consumer);
+
+    auto sqrt = pattern::wrap_type<opset6::Sqrt>({ mean3 }, single_consumer);
+
+    const auto powerOrSqrt = std::make_shared<pattern::op::Or>(OutputVector{ power_sqrt, sqrt });
+
+    // (Sqrt(ReduceMean((x - ReduceMean(x, axes)) ^ 2)) + eps)
```

Comment:
If the difference in this transformations only in `sqrt(...) + eps` and `sqrt(... + eps)` so I would suggest to untie this patterns using `Or`.

Reply:
I thought about that, but it is pretty hard to read those patterns, so I thought that splitting is fine